### PR TITLE
First version of a doc to describe our testing strategy for Vitess.

### PR DESCRIPTION
**Pull request by alainjobart**
Figured it's better if it's written down, as it's not obvious yet in the tree.

### CODE REVIEW

File: test/TestingStrategy.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+## vttest Library, run\_local\_database.py
+
+This set of helpers has two purposes:
+
+* bring up a very light well cluster to help test client applications. YouTube uses this internally with thousands of tests.
+
+* help with Vitess unit tests when a real instance is needed (for instance, to test the interactions with MySQL are what we expect).
+
+It is expected that Vitess users may change py/vttest/environment.py to match their setup. For instance, YouTube internally does is to use different scripts to bring up a MySQL instance.
```
Comment (michael-berlin) [inline test/TestingStrategy.md]: *does this or does it?

File: test/TestingStrategy.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+These tests need to be as light weight as possible, so the developers who use Vitess can run as many unit tests as they want with minimal resources. We are currently running a single MySQL, multiple vtocc, and one vtgate. The plan is to switch to vtcombo to replace vtgate+vtocc processes with a single process. The MySQL process has multiple databases, one per keyspace / shard. It is the smallest setup with a real MySQL.
+
+This framework supports an initial schema to apply to all keyspaces / shards, per keyspace. This is also the base for testing schema changes before rolling them out to production.
+
+Due to its constant nature, this is not an appropriate framework to test cluster events, like re-sharding, re-parenting, advanced query routing, ... No replication is setup on the single MySQL instance, obviously.
+
+## Integration Tests
```
Comment (aaijazi) [inline test/TestingStrategy.md]: I would differentiate between integration tests and end-to-end tests.

Integration tests are things like vtgateconn and tabletconn tests. Anything that, roughly, crosses a single RPC boundary, but doesn't require multiple services to be set up. Arguable the client tests with the fake vtgate are also integration tests.

We do still have (and probably should always have) some e2e tests, such as are in vtgate_v2, resharding tests, etc. They require a working topo server, multiple MySQL instances, etc. 

In the past, we've been a little overzealous with using e2e tests for everything, but we're now doing a lot more integration tests instead.

File: test/TestingStrategy.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+These tests run more complicated setups, and take a lot more resources. They are meant to test end-to-end behaviours of the Vitess echosystem, and complement the unit tests.
```
Comment (michael-berlin) [inline test/TestingStrategy.md]: *behaviors (American English)
*ecosystem

File: test/TestingStrategy.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+* Switch query\_service.py tests to unit tests using vttest. Sugu is on it, first thing is to let vttest only launch a MySQL (and nothing else), and fix its MySQL parameters.
+
+* Switch java tests to use vttest if they need a full cluster, and not java\_vtgate\_test\_helper.py, retire java\_vtgate\_test\_helper.py. Client unit tests should already use vtgateclienttest.
+
+* Same for google3 C++ client.
+
+* We are removing direct vttablet access to python. This in turn will remove a lot of code and tests, like vtclient.py, tablet.py, zkocc.py, ... Less surface area is good, we just need to make sure we maintain good code coverage. As part of this:
+
+  * test/vtdb\_test.py needs to go away. But we need to make sure what it tests is covered in other places. So if it's just testing the python vttablet library, that should go. If it's also testing vttablet timeout behaviours, that should stay.
```
Comment (michael-berlin) [inline test/TestingStrategy.md]: *behaviors

File: test/TestingStrategy.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  * no integration test can depend on direct vttablet access. vtctl has helper query methods for this if needed.
+
+* the python client tests (using vtgateclienttest) are in test/python\_client\_test.py. They need to be finished. Note these are hard to differentiate from the tests in vtgatev2\_test.py. The rule of thumb is if you are testing a client-library feature, tests should be in test/python\_client\_test.py. If you are testing an end-to-end behaviour, tests should be in vtgatev2\_test.py.
```
Comment (michael-berlin) [inline test/TestingStrategy.md]: *behavior

File: test/TestingStrategy.md (hunk 2 above)
Comment (michael-berlin) [inline test/TestingStrategy.md]: I agree with Ammar. In the past, we said integration test, but meant end-to-end test.

The Google Blog has definitions for unit, integration and end-to-end tests: http://googletesting.blogspot.com/2015/04/just-say-no-to-more-end-to-end-tests.html

Maybe it's worthwhile to include short definitions for each test, a good example for such a test and clarify that we aim mostly for unit and integration tests?

File: test/TestingStrategy.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+It is a single stand-alone process, light-weight enough that it runs very quickly and doesn't take many resources at all, and just requires a port to listen on.
+
+(note the go version is run in-process, doesn't even fork vtgateclienttest).
+
+## vttest Library, run\_local\_database.py
+
+This set of helpers has two purposes:
+
+* bring up a very light well cluster to help test client applications. YouTube uses this internally with thousands of tests.
```
Comment (thompsonja) [inline test/TestingStrategy.md]: light weight cluster?

File: test/TestingStrategy.md (hunk 2 above)
Comment (alainjobart) [inline test/TestingStrategy.md]: I agree with name change, these are end to end tests. Will add expectations too.

File: test/TestingStrategy.md (hunk 3 above)
Comment (alainjobart) [inline test/TestingStrategy.md]: echosystem sounds better, better, better, .... better ....
!

**Pull request merged**